lib/jsdom/living/helpers/shadow-dom.js

```diff
@@ -0,0 +1,252 @@
+"use strict";
+
+const NODE_TYPE = require("../node-type");
+const { domSymbolTree } = require("./internal-constants");
+
+const idlUtils = require("../generated/utils");
+
+// Valid host element for ShadowRoot.
+// Defined in: https://dom.spec.whatwg.org/#dom-element-attachshadow
+const VALID_HOST_ELEMENT_NAME = new Set([
+  "article",
+  "aside",
+  "blockquote",
+  "body",
+  "div",
+  "footer",
+  "h1",
+  "h2",
+  "h3",
+  "h4",
+  "h5",
+  "h6",
+  "header",
+  "main",
+  "nav",
+  "p",
+  "section",
+  "span"
+]);
+
+// TODO: Evaluate if it should be somewhere else.
```

Comment:
Importing `../generated/Node` or `../generated/ShadowRoot` would result in a circular dependency. Can we find a good way to avoid this?

Reply:
You could do nodeImpl instanceof NodeImpl, if you import Node-impl.js. That's only slightly better though and probably also has circular dependency issues. This is probably fine.

Reply:
Same issue when importing `Node-impl.js` and `ShadowRoot-impl.js`.